src/codegen.cpp

```diff
@@ -3052,7 +3060,14 @@ static jl_cgval_t emit_call_specfun_other(jl_codectx_t &ctx, jl_code_instance_t
     // emit specialized call site
     jl_value_t *jlretty = codeinst->rettype;
     jl_returninfo_t returninfo = get_specsig_function(jl_Module, specFunctionObject, codeinst->def->specTypes, jlretty);
-    FunctionType *cft = returninfo.decl->getFunctionType();
+    Function *f = returninfo.decl;
+    FunctionType *cft = f->getFunctionType();
+    if (codeinst->def->def.method->pure) {
+        // pure marked functions don't have side-effects, nor observe them
+        f->addFnAttr(Thunk);
+        f->addFnAttr(Attribute::ReadNone);
```

Comment:
Might be good to have a lengthy comment here why `ReadNone` is legal.

Reply:
I'd be more worried about the other one, if I were you. This one is simple: it doesn't read any memory which may be written to, aka ReadNone.

Also, from reading the LICM code, it looks like we should have Attribute::Speculatable also.

Reply:
I still don't think `ReadNone` is correct. Any access to the PTLS is a violation of that and that means the code can't allocate (which is hard to control). While the rules for `@pure` are tricky it took me a while to realize that returning a mutable or a immutable containing an allocated object might not be valid. Even code that allocates, but doesn't return the allocation should then no longer be considered pure, but that code might be of interest to be LICM.

I would like a more fine-grained differentiation between the different kinds of side-effects.